V2: Replace singly linked lists with Set objects

aleclarson opened the pull request.

aleclarson committed:
feat: remove Disposable array support

BREAKING CHANGE

aleclarson committed:
refactor: make `[$addListener]` call itself when passed a `ListenerMap`

To avoid duplicated side effects, `EventEmitter` subclasses that override
the `[$addListener]` method should only apply side effects when a single
function is passed.

BREAKING CHANGE

aleclarson committed:
2.0.0-rc.1

aleclarson committed:
feat: use tslib

Commit:
chore: update lockfile

aleclarson committed:
refactor: remove `EventEmitter.ev` class property

Use the exported `$listeners` symbol, instead.

BREAKING CHANGE

aleclarson committed:
chore: upgrade to typescript@3.9

aleclarson committed:
feat: rewrite using Sets and not singly linked lists

aleclarson committed:
test: update the tests for v2

aleclarson committed:
chore: upgrade tslint

Commit:
chore: update lockfile

aleclarson committed:
fix(types): remove `EventFn` type

aleclarson committed:
fix(types): use any[] for Function type

aleclarson committed:
fix(types): improve `on` return type for single listener

aleclarson committed:
fix: make `off` with 2 arguments always do one thing

aleclarson committed:
fix(types): simplify the `EventKey` type

aleclarson committed:
fix(types): handle special "*" key

aleclarson committed:
chore: update the readme for v2

aleclarson committed:
2.0.0-rc.2

aleclarson committed:
chore: add missing methods to readme

aleclarson committed:
chore: run build step before publish

aleclarson committed:
chore: run tests before publish

aleclarson committed:
fix: remove unused symbol

aleclarson committed:
fix(types): prefer `unknown[]` over `any[]`

aleclarson committed:
fix(types): let `Listener` take any string

aleclarson committed:
chore: add description to `$listeners` symbol

aleclarson committed:
chore: export public api from "src/index.ts"

aleclarson committed:
chore: remove renovate.json

aleclarson committed:
chore: upgrade to prettier@2

aleclarson committed:
chore: remove husky and pretty-quick

aleclarson committed:
chore: remove tslib

Commit:
chore: update lockfile

aleclarson committed:
2.0.0-rc.3

aleclarson committed:
chore: fix `hasListeners` signature in readme

aleclarson committed:
docs: remove semantic-release badge

fearthecowboy commented on src/types.ts:
```diff
@@ -1,33 +1,29 @@
-export type EventIn<T, K extends EventKey<T>> = Id<In<T[K]>>
-export type EventOut<T, K extends EventKey<T>> = Id<Out<T[K]> | void>
+/** Extract an array type of valid event keys */
+export type EventKey<T> = keyof T & string
```
qq: Are you removing support for output values in v2?

aleclarson replied:
Sort of. You'll still be able to use `getListeners` with `Array.from` like this:
```js
const results = Array.from(ee.getListeners('key'), fn => fn())
```
Note: `getListeners` of v2 is not equivalent to `listeners` of v1. The former returns a `Set` of listeners, the latter returns an iterable.

aleclarson committed:
types: split `off` signature into two overloads

aleclarson committed:
feat: change "*" event to "emit"

The "*" key is reserved for removing the listeners of all events, like this:

  ee.on("test", () => {})
  ee.off("*")
  ee.getListeners("test").size // => 0

aleclarson committed:
2.0.0-rc.4

aleclarson committed:
types: add `EventSource` type

To make your EventEmitter readonly, cast it as an EventSource.

aleclarson committed:
2.0.0-rc.5

aleclarson committed:
fix: export EventSource type

aleclarson committed:
2.0.0-rc.6

aleclarson committed:
fix: stop using `Object.freeze` on `emptySet`

You cannot freeze a Set object like that. lol

aleclarson committed:
fix: export `EventSource` with "export type"

or else some bundlers complain

aleclarson committed:
build: use rollup

aleclarson committed:
nit: stop exporting `$listeners` symbol

You can use `ee.getListeners()` instead

aleclarson committed:
types: add object constraint to EventEmitter type parameter

aleclarson committed:
feat(types): let `EventSource` inherit its event types

Pass another `EventSource` or `EventEmitter` type as the type parameter of
the `EventSource` type to use the same events.

Useful for converting an `EventEmitter` to an `EventSource`, as well as
omitting properties not related to events.
